internal/contentmanager/contentmanager.go

```diff
@@ -5,219 +5,331 @@ import (
+		return fmt.Errorf("getting rest.Config: %w", err)
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("getting dynamic client: %w", err)
 	}
 
-	scheme, err := buildScheme(objs)
+	gvkSet, err := gvksForObjects(objs...)
 	if err != nil {
-		return fmt.Errorf("building scheme for ClusterExtension %q: %w", ce.GetName(), err)
+		return fmt.Errorf("getting set of GVKs for managed objects: %w", err)
+	}
+
+	cache, ok := i.caches[ce.Name]
+	if !ok {
+		cache = NewCache()
+	}
+
+	cacheGvks := sets.New[schema.GroupVersionKind]()
+	for gvk := range cache.Sources() {
+		cacheGvks.Insert(gvk)
+	}
+
+	gvksToCreate := gvkSet.Difference(cacheGvks)
+	gvksToRemove := cacheGvks.Difference(gvkSet)
+
+	for _, gvk := range gvksToRemove.UnsortedList() {
+		cache.RemoveSource(gvk)
 	}
 
 	tgtLabels := labels.Set{
 		oclabels.OwnerKindKey: v1alpha1.ClusterExtensionKind,
 		oclabels.OwnerNameKey: ce.GetName(),
 	}
 
-	c, err := cache.New(cfg, cache.Options{
-		Scheme:               scheme,
-		DefaultLabelSelector: tgtLabels.AsSelector(),
+	sharedInfFact := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynamicClient, time.Hour*10, metav1.NamespaceAll, func(lo *metav1.ListOptions) {
+		lo.LabelSelector = tgtLabels.AsSelector().String()
 	})
+
+	scheme, err := buildScheme(gvksToCreate.UnsortedList()...)
 	if err != nil {
-		return fmt.Errorf("creating cache for ClusterExtension %q: %w", ce.Name, err)
+		return fmt.Errorf("building scheme: %w", err)
 	}
 
-	for _, obj := range objs {
-		err = ctrl.Watch(
-			source.Kind(
-				c,
-				obj,
-				handler.TypedEnqueueRequestForOwner[client.Object](
-					scheme,
-					i.mapper,
-					ce,
-					handler.OnlyControllerOwner(),
-				),
+	sourcesErrors := []error{}
+	createdSourcesMap := map[schema.GroupVersionKind]ReaderCloserSyncingSource{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		restMapping, err := i.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			sourcesErrors = append(sourcesErrors, fmt.Errorf("getting resource mapping for GVK %q: %w", gvk, err))
+			continue
+		}
+
+		s := &DynamicInformerSource{
+			GVR:     restMapping.Resource,
+			Owner:   ce,
+			Handler: handler.EnqueueRequestForOwner(scheme, i.mapper, ce, handler.OnlyControllerOwner()),
+			Predicates: []predicate.Predicate{
 				predicate.Funcs{
 					CreateFunc:  func(tce event.TypedCreateEvent[client.Object]) bool { return false },
 					UpdateFunc:  func(tue event.TypedUpdateEvent[client.Object]) bool { return true },
 					DeleteFunc:  func(tde event.TypedDeleteEvent[client.Object]) bool { return true },
 					GenericFunc: func(tge event.TypedGenericEvent[client.Object]) bool { return true },
 				},
-			),
-		)
+			},
+			DynamicInformerFactory: sharedInfFact,
+			OnWatchError: func() {
+				i.mu.Lock()
+				defer i.mu.Unlock()
+				if cache, ok := i.caches[ce.Name]; ok {
+					cache.RemoveSource(gvk)
+				}
+			},
```

Comment:
I _think_ this works, but I haven't been able to find a way to test this logic. I'm open to suggestions on ways I could try to force this functionality to execute in a live cluster.

If there is no way we can think of to force this behavior to execute, maybe exercising this behavior alone is worth taking the time to implement unit tests for this method.